File: cms-kyc-customer/src/main/java/com/github/knextsunj/cms/cmskyccustomer/service/impl/CustomerServiceImpl.java
package com.github.knextsunj.cms.cmskyccustomer.service.impl;


import com.github.knextsunj.cms.cmskyccustomer.exception.BusinessException;
import com.github.knextsunj.cms.cmskyccustomer.mapper.CustomerMapper;
import com.github.knextsunj.cms.cmskyccustomer.repository.CustomerRepository;
import com.github.knextsunj.cms.cmskyccustomer.service.CustomerService;
import com.github.knextsunj.cmskyccustomer.domain.Customer;
import com.github.knextsunj.cmskyccustomer.dto.CustomerDTO;
import io.github.resilience4j.bulkhead.annotation.Bulkhead;
import io.github.resilience4j.retry.RetryRegistry;
import io.github.resilience4j.retry.annotation.Retry;
import org.apache.commons.lang3.tuple.Pair;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.InitializingBean;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

@Service
public class CustomerServiceImpl implements CustomerService, InitializingBean {

    private static final Logger logger = LoggerFactory.getLogger(CustomerServiceImpl.class);

    @Autowired
    private CustomerMapper customerMapper;

    @Autowired
    private CustomerRepository customerRepository;

    @Autowired
    private RetryRegistry retryRegistry;

    @Override
    @Retry(name="saveCustomerForKyc",fallbackMethod = "saveCustomerFallback")
    public boolean saveCustomer(CustomerDTO customerDTO) {
        logger.debug("Saving data for customer kyc: {}",customerDTO);
        Customer customer = customerRepository.insert(customerMapper.fromCustomerDTO(customerDTO));
        return customer != null ? true : false;
    }

    @Override
    @Retry(name="updateCustomerKyc",fallbackMethod = "updateCustomerFallback")
    public Pair updateCustomer(CustomerDTO customerDTO) {

        boolean result = false;

        if (customerDTO.getDeleted().equals("Y")) {
            Optional<Customer> customerOptional = customerRepository.findById(customerDTO.getId());
            if (customerOptional.isPresent()) {
                try {
                    Customer customer = customerOptional.get();
                    customer.setDeleted("Y");
                    customer = customerRepository.save(customer);
                    result = customer != null ? true : false;
                } catch (Exception ex) {
                    logger.error("Failed to delete customer", ex);
                    throw new BusinessException("Failed to delete customer",ex);
                }
            }
        }
        Customer customer = customerRepository.save(customerMapper.fromCustomerDTO(customerDTO));
        result = customer != null ? true : false;
        return Pair.of(customerDTO,result);
    }

    @Override
    @Bulkhead(name="findAllCustomerBulkhead",fallbackMethod = "findAllCustomersFallback")
    public List<CustomerDTO> findAllCustomers() {
        return customerRepository.findAllByCustomerStatusAndDeleted("ACTIVE", "N").stream().map(customer -> customerMapper.toCustomerDTO(customer)).collect(Collectors.toList());
    }

    @Override
    public List<CustomerDTO> findAllDeletedCustomers() {
        return customerRepository.findAllByDeleted("Y").stream().map(customer -> customerMapper.toCustomerDTO(customer)).collect(Collectors.toList());
    }

    private boolean saveCustomerFallback(CustomerDTO customerDTO, RuntimeException runtimeException) {
        return false;
    }

    private Pair updateCustomerFallback(CustomerDTO customerDTO, RuntimeException runtimeException) {
        return Pair.of(customerDTO,false);
    }

    private List<CustomerDTO> findAllCustomersFallback() {
        return Collections.emptyList();
    }

    @Override
    public void afterPropertiesSet() throws Exception {
        retryRegistry.retry("saveCustomerForKyc")
                .getEventPublisher()
                .onRetry(System.out::println);

        retryRegistry.retry("updateCustomerKyc")
                .getEventPublisher()
                .onRetry(System.out::println);
    }


}

File: cms-kyc/src/main/kotlin/com/github/knextsunj/cms/cmskyc/service/impl/CustomerServiceImpl.kt
package com.github.knextsunj.cms.cmskyc.service.impl

import com.fasterxml.jackson.databind.ObjectMapper
import com.github.knextsunj.cms.cmskyc.dto.AddressDTO
import com.github.knextsunj.cms.cmskyc.dto.CustomerDTO
import com.github.knextsunj.cms.cmskyc.service.BaseServiceImpl
import com.github.knextsunj.cms.cmskyc.service.CustomerService
import io.github.resilience4j.circuitbreaker.annotation.CircuitBreaker
import io.github.resilience4j.ratelimiter.RequestNotPermitted
import io.github.resilience4j.ratelimiter.annotation.RateLimiter
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.core.env.Environment
import org.springframework.http.HttpEntity
import org.springframework.http.HttpHeaders
import org.springframework.http.HttpMethod
import org.springframework.http.ResponseEntity
import org.springframework.stereotype.Service
import org.springframework.web.client.RestTemplate
import java.time.LocalDateTime

@Service
open class CustomerServiceImpl : CustomerService, BaseServiceImpl() {

    companion object {
        private val logger: Logger = LoggerFactory.getLogger(CustomerServiceImpl::class.java)
    }

    @Autowired
    open lateinit var environment: Environment

    @Autowired
    open lateinit var objectMapper: ObjectMapper

    @Autowired
    open lateinit var restTemplate: RestTemplate

    /**
     * Computed property using parent abstract java class field as backing field
     */
    private val customerBaseUrl:String
        get():String{
           baseCustomerUrl  = environment.getProperty("cms-kyc.customer.base.url")
            return baseCustomerUrl
        }


    @CircuitBreaker(name = "saveOrUpdateCustomerCallCircuitBreaker", fallbackMethod = "logSaveOrUpdateCustomerCallFailure")
    override fun saveCustomer(customerDTO: CustomerDTO?,authHeader:String): Boolean {
        val fullUrl = customerBaseUrl+"/save"
//        val url = environment.get("cms-kyc-customer.save.url")
        val headers = HttpHeaders()
        headers.set(HttpHeaders.AUTHORIZATION,authHeader)
        val entity = HttpEntity(customerDTO, headers)
        logger.info("Calling customer kyc microservice for posting data: {}",customerDTO)
        logger.info("Current time is::: {}",LocalDateTime.now());
        return restTemplate.postForObject(fullUrl, entity, Boolean::class.java)!!
    }

    @CircuitBreaker(name = "saveOrUpdateCustomerCallCircuitBreaker", fallbackMethod = "logSaveOrUpdateCustomerCallFailure")
    override fun updateCustomer(customerDTO: CustomerDTO?,authHeader:String): Boolean {
        val fullUrl = customerBaseUrl+"/update"
//        val url = environment.get("cms-kyc-customer.update.url")
        val headers = HttpHeaders()
        headers.set(HttpHeaders.AUTHORIZATION,authHeader)
        val request: HttpEntity<CustomerDTO> = HttpEntity(customerDTO,headers)
        val response: ResponseEntity<Boolean> = restTemplate.exchange(fullUrl, HttpMethod.PUT, request, Boolean::class.java)
        return response.body
    }

    override fun findAllCustomers(authHeader:String): MutableList<CustomerDTO> {
//        val url: String? = environment.get("cms-kyc-customer.findall.url")
        val fullUrl = customerBaseUrl+"/fetchAll"
        val headers = HttpHeaders()
        headers.set(HttpHeaders.AUTHORIZATION,authHeader)
        val request = HttpEntity(null,headers)
        val responseEntity = restTemplate.getForEntity(fullUrl, MutableList::class.java,request)
        return responseEntity.body as MutableList<CustomerDTO>
    }

    @RateLimiter(name = "findAllDeletedCustomersRateLimiter", fallbackMethod = "findAllDeletedCustomersFallback")
    override fun findAllDeletedCustomers(authHeader:String): List<CustomerDTO> {
        val fullUrl = customerBaseUrl+"/fetchAllDeleted"
        val headers = HttpHeaders()
        headers.set(HttpHeaders.AUTHORIZATION,authHeader)
        val request = HttpEntity(null,headers)
        val responseEntity = restTemplate.getForEntity(fullUrl, MutableList::class.java,request)
        return responseEntity.body as MutableList<CustomerDTO>
    }

    private fun findAllDeletedCustomersFallback(rnp: RequestNotPermitted):List<CustomerDTO> {
        logger.error("Timeout exception when fetching all deleted customers",rnp);
        return emptyList()
    }

    private fun logSaveOrUpdateCustomerCallFailure(customerDTO: CustomerDTO?, authHeader:String, exception: RuntimeException):Boolean {
//       logger.error("Failed to save or update address ",exception)
        logger.error("logSaveOrUpdateCustomerCallFailure called")
        return false
    }
}